js_modules/dagster-ui/packages/ui-core/src/overview/assets/AssetTimelineDataFetcher.tsx

```diff
@@ -0,0 +1,414 @@
+      const currentKeys = Object.keys(current[0]);
+      return currentKeys.every((key) => prevKeys.has(key));
+    },
+    [codeLocationNodes, didRangeMeaningfullyChange] as const,
+  );
+
+  const client = useApolloClient();
+  React.useEffect(() => {
+    let isInProgress = false;
+    let isCanceled = false;
+    async function tick(isRefresh: boolean = false) {
+      if (isInProgress) {
+        return;
+      }
+      isInProgress = true;
+      await Promise.allSettled(
+        Object.keys(codeLocationSubscriptions.current).map((locationName) => {
+          const node = codeLocationNodes[locationName];
+          const groups = node?.groups;
+          if (!groups) {
+            return Promise.resolve();
+          }
+          const assetToGroup: Record<string, string> = {};
+          const assetKeys = Object.values(groups)
+            .flatMap(({assets, groupId}) => {
+              assets.forEach((asset) => {
+                const stringKey = tokenForAssetKey(asset.key);
+                assetToGroup[stringKey] = groupId;
+              });
+              return assets;
+            })
+            .map((asset) => ({path: asset.key.path}));
+
+          if (!assetKeys.length) {
+            return Promise.resolve();
+          }
+
+          function triggerListeners() {
+            const listener = codeLocationSubscriptions.current[locationName];
+            if (listener) {
+              listener(runsCache.current.locations[locationName]!);
+            }
+
+            Object.keys(groups!).forEach((groupKey) => {
+              const listener = groupSubscriptions.current[groupKey];
+              if (listener) {
+                listener(runsCache.current.groups[groupKey]!);
+              }
+              assetKeys.forEach((key) => {
+                const stringAssetKey = tokenForAssetKey(key);
+                const listener = assetSubscriptions.current[stringAssetKey];
+                if (listener) {
+                  listener(runsCache.current.assets[stringAssetKey]!);
+                }
+              });
+            });
+          }
+
+          if (isRefresh) {
+            // Mark all of the assets as loading again
+            runsCache.current.locations[locationName] = {
+              runs: runsCache.current.locations[locationName]?.runs ?? [],
+              loading: true,
+            };
+            Object.keys(groups).forEach((groupKey) => {
+              runsCache.current.groups[groupKey] = {
+                runs: runsCache.current.groups[groupKey]?.runs ?? [],
+                loading: true,
+              };
+              assetKeys.forEach((key) => {
+                const stringAssetKey = tokenForAssetKey(key);
+                runsCache.current.assets[stringAssetKey] = {
+                  runs: runsCache.current.assets[stringAssetKey]?.runs ?? [],
+                  loading: true,
+                };
+              });
+            });
```

Comment:
This `tick` function is quite a beast. I can't think of good ways to break it up, but I think there is some super important logic in here that would be easier to grok if we pushed some of these things into named helper methods.

Something like `markAllAsLoading(runsCache, locationName, groups, assetKeys)` would be super nice here